### PR TITLE
FRR: Load VRF kernel module upon initial startup, disable use of vrfs when not available

### PR DESCRIPTION
**Pull request by jbemmel**
Fixes https://github.com/ipspace/netlab/issues/1209

### CODE REVIEW

File: netsim/ansible/tasks/frr/initial-clab.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    - name: "Load VRF kernel module"
+      when: netlab_mgmt_vrf|default(False) or 'vrf' in module
+      shell: |
+        modprobe vrf
+      become: true
+      delegate_to: localhost
+      run_once: true
+      tags: [ print_action, always ]
+  rescue:
+    - name: Unable to load vrf kernel module - disable use of VRFs
```
Comment (jbemmel) [inline netsim/ansible/tasks/frr/initial-clab.yml:12]: Could report an error if 'vrf' module is being used

**Comment by ipspace** [comment]
Love the idea. However:

* Ideally, we'd fix this in the FRR quirks. 
* **modprobe** can be executed as non-root but will fail if it needs to load the driver. As the **vrf** driver is loaded if installed, this shouldn't be a big dea.
* I would deal with the VRF module separately (in the VRF configuration task list). 

Finally:

* **run_once** is implemented by executing the task on a single node, which means that the **netlab_mgmt_vrf** would be set only for that node.

So, I would propose to add the "modprobe" call to FRR quirks and act on that. It would also be easier to cache the results there (ensuring we only call **modprobe** once). I could add that if you wish (but not before tomorrow or so).

**Comment by ipspace** [comment]
... also, if we just want to test whether the VRF module is loaded, looking for 'vrf' in **lsmod** printout is a better alternative.

**Comment by jbemmel** [comment]
> Finally:
> 
> * **run_once** is implemented by executing the task on a single node, which means that the **netlab_mgmt_vrf** would be set only for that node.
> 

I tested for that, actually. The ```modprobe``` runs once, but the failure handler runs for every frr node :)

image

**Comment by jbemmel** [comment]
> ... also, if we just want to test whether the VRF module is loaded, looking for 'vrf' in **lsmod** printout is a better alternative.

I checked that the vrf kernel module is loaded automatically on demand when one issues the ```ip link ... type vrf``` command. So it not being loaded doesn't mean that it won't work

**Comment by ipspace** [comment]
> > * **run_once** is implemented by executing the task on a single node, which means that the **netlab_mgmt_vrf** would be set only for that node.
> 
> I tested for that, actually. The `modprobe` runs once, but the failure handler runs for every frr node :)

That's REALLY NICE to know (thanks a million), I'm just worried that the behavior might change in another Ansible release. 

Anyway, should we go down this path or implement a quirk? Using a quirk would give us more control earlier in the process, and the ability to generate a warning.

**Comment by jbemmel** [comment]
> Anyway, should we go down this path or implement a quirk? Using a quirk would give us more control earlier in the process, and the ability to generate a warning.

I modeled it after what we're doing for the mpls kernel modules, I think this is similar and consistent. I do agree that the vrf module case is different and should be checked for and aborted earlier in the process, with a decent warning

The 'initial' config can work with or without management vrf

**Comment by ipspace** [comment]
> > Anyway, should we go down this path or implement a quirk? Using a quirk would give us more control earlier in the process, and the ability to generate a warning.
> 
> I modeled it after what we're doing for the mpls kernel modules, I think this is similar and consistent.

To be fair, we probably added the MPLS hack before implementing device quirks ;)

> I do agree that the vrf module case is different and should be checked for and aborted earlier in the process, with a decent warning

We have to be consistent and do one or the other for the VRF and the MPLS modules.

> The 'initial' config can work with or without management vrf

... but it would be nice to know that we turned the management VRF off ;)

Anyway, will merge this one as-is and then work on a more generic solution. Thank you!

**Pull request merged**